Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js

```diff
@@ -1664,12 +1672,8 @@ WI.DebuggerManager = class DebuggerManager extends WI.Object
     {
         this._ignoreBreakpointDisplayLocationDidChangeEvent = true;
 
-        for (let breakpoint of breakpoints) {
-            if (!breakpoint.sourceCodeLocation.sourceCode)
-                breakpoint.sourceCodeLocation.sourceCode = sourceCode;
-            // SourceCodes can be unequal if the SourceCodeLocation is associated with a Script and we are looking at the Resource.
-            console.assert(breakpoint.sourceCodeLocation.sourceCode === sourceCode || breakpoint.sourceCodeLocation.sourceCode.contentIdentifier === sourceCode.contentIdentifier);
-        }
+        for (let breakpoint of breakpoints)
+            breakpoint.sourceCodeLocation.sourceCode = sourceCode;
```

Comment:
i thought about this a bit more and realized this could be problematic if the same script is loaded more than once (e.g. once in the main page and once in an `<iframe>`) since the `contentIdentifier` is really just an alias for the URL/`sourceURL`

i can see why we do need to do something like this, but it probably should somehow be gated to "once per top-level navigation" or something like that

i admit this is probably a pretty rare occurrence, so it's maybe fine to do as a followup?  at the very least i think we should verify whether my concern is actually valid and if so write a test with a FIXME comment pointing to an open bug

Reply:
yeah we'd see the same resource (at least something with the same URL) appear twice

i think the only way this would really be concerning is if somehow the second load returned different content than the first (which is totally possible, but probably not likely)

currently we map breakpoints to the first resource loaded with the given URL, but i think with this change we might end up mapping to the most recently loaded resource instead

again this is probably pretty uncommon so im not too concerned about the difference in behavior (and even then the contents of both resources are likely the same, so we'd just be showing different hierarchical path component navigation items)

maybe we could do some really basic check in the meantime that only does this if we're not in the current session?  something like
```js
let resource = sourceCode;
if (resource instanceof WI.Script)
    resource = resource.resource;
console.assert(!resource || resource instanceof WI.Resource, resource);
for (let parentFrame = resource?.parentFrame; parentFrame; parentFrame = parentFrame.parentFrame) {
    if (parentFrame === WI.networkManager.mainFrame)
        return;
}
```

Reply:
> currently we map breakpoints to the first resource loaded with the given URL, but i think with this change we might end up mapping to the most recently loaded resource instead
> 
> ...
> 
> maybe we could do some really basic check in the meantime that only does this if we're not in the current session? something like
> 
> ```js
> let resource = sourceCode;
> if (resource instanceof WI.Script)
>     resource = resource.resource;
> console.assert(!resource || resource instanceof WI.Resource, resource);
> for (let parentFrame = resource?.parentFrame; parentFrame; parentFrame = parentFrame.parentFrame) {
>     if (parentFrame === WI.networkManager.mainFrame)
>         return;
> }
> ```

Wait, I thought the point here is we want to ensure that we still only set the source code if the source code came from the main frame / first loaded frame. The snippet however looks like it's skipping setting the source code if it indeed came from the main frame, which is the opposite of what we want? What am I missing 😅

Reply:
oh oops yes i wasn't entirely clear.  my apologies 😅

what i meant was that we should only set `breakpoint.sourceCodeLocation.sourceCode = sourceCode;` if the current value for `breakpoint.sourceCodeLocation.sourceCode` is either `null` or not for the main frame.  this way, we'd preserve the existing behavior of only setting the `sourceCode` once (for a given main frame) while still allowing for the new behavior of it changing as the user navigates/refreshes (i.e. across main frames)

so putting it all together id imagine adding this
```js
_isActiveSourceCode(sourceCode)
{
    let resource = sourceCode;

    if (sourceCode instanceof WI.Script) {
        resource = resource.resource;
        if (!resource)
            return this.knownNonResourceScripts.includes(sourceCode);
    }

    console.assert(!resource || resource instanceof WI.Resource, resource);

    for (let parentFrame = resource?.parentFrame; parentFrame; parentFrame = parentFrame.parentFrame) {
        if (parentFrame === WI.networkManager.mainFrame)
            return true;
    }
    return false;
}

_associateBreakpointsWithSourceCode(breakpoints, sourceCode)
{
    if (!this._isActiveSourceCode(sourceCode))
        return;

    this._ignoreBreakpointDisplayLocationDidChangeEvent = true;

    for (let breakpoint of breakpoints) {
        if (this._isActiveSourceCode(breakpoint.sourceCodeLocation.sourceCode))
            continue;

        breakpoint.sourceCodeLocation.sourceCode = sourceCode;
    }

    this._ignoreBreakpointDisplayLocationDidChangeEvent = false;
}
```

im pretty sure this should work even with something like a `JSContext`/`ServiceWorker`/etc. where there are only `WI.Script` without any associated `WI.Resource`, but id suggest double checking just to be sure :)

this is pretty complex, however, so like i mentioned before i think it's fine to have the basic fix for the (likely) common case be done in this PR and the above fix for the (likely) uncommon case be done in a followup

Reply:
I filed https://bugs.webkit.org/show_bug.cgi?id=274914 for this potential follow-up fix.